_posts/2024-04-12-learn-from-the-xz-backdoor.md

```diff
@@ -0,0 +1,111 @@
+
+So the SUSE security team and myself, we started analyzing. With the initial information that there is something suspicious, it was relatively easy to find more suspicious things, in no particular order:
+
+* openSUSE and SUSE is tracking release artifact signatures with a keyring of trusted signatures. We noticed that the key that the artifacts were signed with changed some time ago, so we had to update our trusted keyring for the XZ project. We validated that there was a maintainer handover and that the new maintainer has direct commit access as well as the ability to sign releases and publish them. The web of trust of this new signing key was not well connected, which could have raised an alert, but it was signed by the previous maintainer and that was sufficient for us.
+
+* Looking at the commit history, there was a flurry of commits between the last 5.5 beta and the 5.6.0 release in a short time window, by only one committer, not coming via a PR and no obvious review or discussion on it. This was immediately concerning. Normally projects don't do that just before a major new release. Reviewing every single commit immediately showed odd test files being committed and updated in 5.6.1, that did not have corresponding updates in the test framework or in the project code, so were "unused". Normally test files are committed alongside a code fix in the same commit or with a reference to a prior issue or commit that this test case is addressing. For an experienced maintainer of an upstream project, this seemed like a big oversight. The commit messages were sort of plausible but not really making sense, especially when comparing the (small) differences between 5.6.0 and 5.6.1.
+
+Further investigation lead to finding the "stage zero" embedded in the build system and with that we were able to step through the layers of obfuscation to untangle the 2nd and 3rd stage. Within minutes it became clear to us that very significant effort was spent on developing it. It wasn't the work of a single developer on a rainy sunday afternoon. Also, the 2nd stage hinted that this was a backdoor that was specifically targetting only specific environments, Debian or RPM package builds using GCC and glibc. A normal user building from source, either the backdoored tarball or from git would have never been affected. This raised the alarm bells. So before we went further with the reverse engineering, we assessed the impact.
+
```

Comment:
I would add:
+ One day before disclosure, on Thursday evening, SUSE product security received a longer and detailed report from Andres Freund via the shared distros security disclosure list. The distros list is an encrypted mailing list where distributors collaborate and coordinate on coordinated disclosures of security issues. This report brought the new knowledge that the xz backdoor specifically targeting openssh, which is one of the network facing parts of nearly every Linux system. This even further increased our threatlevel to be of a remote access backdoor and caused also to widen our planned communication efforts.
+ SUSE product security is a member of various private security forums, like the distros list and CERT VINCE and others, which allow us the coordinate fixes between software vendors and have updates ready on the disclosure dates.